### PR TITLE
Allow to adjust read and write timeouts after connecting

### PR DESCRIPTION
**Pull request by casperisfine**
That's a feature of `mysql2` we use often.

### CODE REVIEW

File: contrib/ruby/test/client_test.rb [hunk 1]
```diff
@@ -264,6 +264,55 @@ def test_read_timeout
...
+  end
+
+  def test_read_timeout_closed_connection
+    client = new_tcp_client(read_timeout: 5)
+    client.close
+    ensure_closed client
+
+    assert_raises IOError do
+      client.read_timeout
+    end
```
Comment (casperisfine) [inline contrib/ruby/test/client_test.rb:286]: So this is a bit weird, but seemed the most logical.

Alternatively I could return a fallback value, or I could keep a copy of the timeout value in the client object, and return that.

**Review by composerinteralia — APPROVED** [review]
Looks good to me. I think erroring once the connection is closed makes sense, unless there is a specific reason to do otherwise.

One rounding quirk that I don't think was caused by anything in this PR but is more obvious now:

Screen Shot 2022-01-18 at 4 17 57 PM

**Comment by casperisfine** [comment]
> I think erroring once the connection is closed makes sense, unless there is a specific reason to do otherwise.

Yeah, I'm not quite sure at this stage. I suppose it's not really expected, so maybe it would make more sense to simply noop?

It's likely that the calling code looks something like:

```ruby
client.read_timeout = something
client.query(...)
```

People might not expect `read_timeout=` to raise (even less so IOError) and only rescue around `query`.

> One rounding quirk

Yeah it's stored as a double, so this is bound to happen, unless we keep the original `VALUE` inside the Ruby object.

**Comment by composerinteralia** [comment]
> Yeah, I'm not quite sure at this stage. I suppose it's not really expected, so maybe it would make more sense to simply noop?

I'll merge as is for now. We can always come back and change it if needed.

**Pull request merged**